File: plugins/plugin-instagram/rust/src/actions/post_comment.rs
//! Post comment action for Instagram

use async_trait::async_trait;
use serde_json::Value;

use super::{ActionContext, InstagramAction};
use crate::error::Result;

/// Action to post a comment on Instagram media
pub struct PostCommentAction;

#[async_trait]
impl InstagramAction for PostCommentAction {
    fn name(&self) -> &'static str {
        "POST_INSTAGRAM_COMMENT"
    }

    fn description(&self) -> &'static str {
        "Post a comment on an Instagram post or media"
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        // Check if this is an Instagram message with a media ID
        let source = context.message.get("source").and_then(|v| v.as_str());
        let has_media = context.media_id.is_some();
        Ok(source == Some("instagram") && has_media)
    }

    async fn execute(&self, context: &ActionContext) -> Result<Value> {
        let response_text = context
            .state
            .get("response")
            .and_then(|v| v.get("text"))
            .and_then(|v| v.as_str())
            .or_else(|| context.message.get("text").and_then(|v| v.as_str()))
            .unwrap_or("");

        Ok(serde_json::json!({
            "action": self.name(),
            "media_id": context.media_id,
            "text": response_text
        }))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_validate_instagram_source_with_media() {
        let action = PostCommentAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "instagram",
                "text": "Great post!"
            }),
            user_id: 12345,
            thread_id: None,
            media_id: Some(67890),
            state: serde_json::json!({}),
        };

        assert!(action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_validate_non_instagram_source() {
        let action = PostCommentAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "telegram",
                "text": "Great post!"
            }),
            user_id: 12345,
            thread_id: None,
            media_id: Some(67890),
            state: serde_json::json!({}),
        };

        assert!(!action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_validate_no_media() {
        let action = PostCommentAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "instagram",
                "text": "Great post!"
            }),
            user_id: 12345,
            thread_id: None,
            media_id: None,
            state: serde_json::json!({}),
        };

        assert!(!action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_execute_with_response() {
        let action = PostCommentAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "instagram",
                "text": "fallback text"
            }),
            user_id: 12345,
            thread_id: None,
            media_id: Some(67890),
            state: serde_json::json!({
                "response": {
                    "text": "This is a great post!"
                }
            }),
        };

        let result = action.execute(&context).await.unwrap();
        assert_eq!(result["action"], "POST_INSTAGRAM_COMMENT");
        assert_eq!(result["media_id"], 67890);
        assert_eq!(result["text"], "This is a great post!");
    }

    #[tokio::test]
    async fn test_execute_with_fallback_text() {
        let action = PostCommentAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "instagram",
                "text": "Fallback comment"
            }),
            user_id: 12345,
            thread_id: None,
            media_id: Some(67890),
            state: serde_json::json!({}),
        };

        let result = action.execute(&context).await.unwrap();
        assert_eq!(result["text"], "Fallback comment");
    }
}

File: plugins/plugin-instagram/typescript/src/actions/post-comment.ts
/**
 * Post comment action for Instagram
 */

import type {
  Action,
  ActionExample,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { INSTAGRAM_SERVICE_NAME } from "../constants";
import type { InstagramService } from "../service";

/**
 * Action to post a comment on Instagram media
 */
export const postCommentAction: Action = {
  name: "POST_INSTAGRAM_COMMENT",
  description: "Post a comment on an Instagram post or media",
  similes: [
    "instagram_comment",
    "comment_instagram",
    "reply_instagram",
    "post_comment_instagram",
  ],
  examples: [
    [
      {
        name: "{{user1}}",
        content: {
          text: "Comment on this Instagram post",
        },
      },
      {
        name: "{{agent}}",
        content: {
          text: "I'll post a comment on that Instagram post.",
          action: "POST_INSTAGRAM_COMMENT",
        },
      },
    ],
  ] as ActionExample[][],

  async validate(
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
  ): Promise<boolean> {
    // Check if this is an Instagram context with a media ID
    const content = message.content as Record<string, unknown>;
    const source = content.source as string | undefined;
    const mediaId = content.mediaId as number | undefined;

    if (source !== "instagram") {
      return false;
    }

    if (!mediaId) {
      return false;
    }

    // Check if Instagram service is available
    const service = runtime.getService(INSTAGRAM_SERVICE_NAME);
    if (!service) {
      return false;
    }

    return true;
  },

  async handler(
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
  ) {
    const service = runtime.getService(
      INSTAGRAM_SERVICE_NAME,
    ) as InstagramService;
    if (!service || !service.getIsRunning()) {
      if (callback) {
        await callback({
          text: "Instagram service is not running",
        });
      }
      return { success: false, error: "Service not available" };
    }

    const content = message.content as Record<string, unknown>;
    const mediaId = content.mediaId as number | undefined;

    if (!mediaId) {
      if (callback) {
        await callback({
          text: "No media ID provided for Instagram comment",
        });
      }
      return { success: false, error: "Missing media ID" };
    }

    // Get response text from state or message
    const responseText =
      ((state?.response as Record<string, unknown> | undefined)?.text as
        | string
        | undefined) ||
      (content.text as string | undefined) ||
      "";

    if (!responseText) {
      if (callback) {
        await callback({
          text: "No comment text to post",
        });
      }
      return { success: false, error: "Empty comment" };
    }

    try {
      const commentId = await service.postComment(mediaId, responseText);

      if (callback) {
        await callback({
          text: `Comment posted on Instagram media ${mediaId}`,
          action: "POST_INSTAGRAM_COMMENT",
        });
      }

      return { success: true, data: { mediaId, commentId } };
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : "Unknown error";
      if (callback) {
        await callback({
          text: `Failed to post Instagram comment: ${errorMessage}`,
        });
      }
      return { success: false, error: errorMessage };
    }
  },
};
